test/integration/apiserver/admissionwebhook/load_balance_test.go

```diff
@@ -48,6 +50,19 @@ const (
 
 // TestWebhookLoadBalance ensures that the admission webhook opens multiple connections to backends to satisfy concurrent requests
 func TestWebhookLoadBalance(t *testing.T) {
+	srv, err := mockdns.NewServer(map[string]mockdns.Zone{
+		"example.com.": {
+			A:    []string{"127.0.0.1"},
+			AAAA: []string{"::1"},
+		},
+	}, false)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer srv.Close()
+
+	srv.PatchNet(net.DefaultResolver)
+	defer mockdns.UnpatchNet(net.DefaultResolver)
```

Comment:
oh, smart idea, I have a package for just that, I also think I can reduce it to a file we can add in the test folder https://github.com/aojea/mem-resolver , see the https://github.com/aojea/mem-resolver/blob/main/example_test.go

Reply:
this seems to modify the global resolver 🤔

Reply:
there is an easier way, we can override the Service Resolver, I left you the exercise of generating the certificate to work with the Service name

```diff
diff --git a/test/integration/apiserver/admissionwebhook/load_balance_test.go b/test/integration/apiserver/admissionwebhook/load_balance_test.go
index 9423b0f0dc1..38df89c056f 100644
--- a/test/integration/apiserver/admissionwebhook/load_balance_test.go
+++ b/test/integration/apiserver/admissionwebhook/load_balance_test.go
@@ -25,6 +25,7 @@ import (
        "io"
        "net"
        "net/http"
+       "net/url"
        "sync"
        "sync/atomic"
        "testing"
@@ -38,6 +39,7 @@ import (
        "k8s.io/apimachinery/pkg/util/wait"
        clientset "k8s.io/client-go/kubernetes"
        "k8s.io/client-go/rest"
+       "k8s.io/kubernetes/cmd/kube-apiserver/app"
        kubeapiservertesting "k8s.io/kubernetes/cmd/kube-apiserver/app/testing"
        "k8s.io/kubernetes/test/integration/framework"
 )
@@ -46,6 +48,12 @@ const (
        testLoadBalanceClientUsername = "webhook-balance-integration-client"
 )
 
+type staticURLServiceResolver string
+
+func (u staticURLServiceResolver) ResolveEndpoint(namespace, name string, port int32) (*url.URL, error) {
+       return url.Parse(string(u))
+}
+
 // TestWebhookLoadBalance ensures that the admission webhook opens multiple connections to backends to satisfy concurrent requests
 func TestWebhookLoadBalance(t *testing.T) {
 
@@ -80,7 +88,12 @@ func TestWebhookLoadBalance(t *testing.T) {
        }()
        defer httpServer.Close()
 
-       webhookURL := "https://" + localListener.Addr().String()
+       _, port, err := net.SplitHostPort(localListener.Addr().String())
+       if err != nil {
+               t.Fatalf("unexpected error: %v", err)
+       }
+
+       t.Cleanup(app.SetServiceResolverForTests(staticURLServiceResolver(fmt.Sprintf("https://127.0.0.1:%s", port))))
 
        s := kubeapiservertesting.StartTestServerOrDie(t, kubeapiservertesting.NewDefaultTestServerOptions(), []string{
                "--disable-admission-plugins=ServiceAccount",
@@ -119,7 +132,10 @@ func TestWebhookLoadBalance(t *testing.T) {
                Webhooks: []admissionregistrationv1.MutatingWebhook{{
                        Name: "admission.integration.test",
                        ClientConfig: admissionregistrationv1.WebhookClientConfig{
-                               URL:      &webhookURL,
+                               Service: &admissionregistrationv1.ServiceReference{
+                                       Namespace: "test",
+                                       Name:      "webhook",
+                               },
                                CABundle: localhostCert,
                        },
                        Rules: []admissionregistrationv1.RuleWithOperations{{
```

we should add integration test for http2 and http1

Reply:
> this seems to modify the global resolver

I am annoyed that resolver is not an interface, which would make mocking much easier.

> we should add integration test for http2 and http1

Added.